library/core/src/hint.rs

```diff
@@ -121,7 +121,30 @@ pub fn black_box<T>(dummy: T) -> T {
     #[cfg(not(miri))] // This is just a hint, so it is fine to skip in Miri.
     // SAFETY: the inline assembly is a no-op.
     unsafe {
-        llvm_asm!("" : : "r"(&dummy));
+        #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+        {
+            asm!(
+                "/* {0} */",
+                in(reg) &dummy,
```

Comment:
I think we probably want `&mut dummy` here to indicate to the compiler that `dummy` may be modified for optimization purposes (even if it isn't). The current code still works but it's a bit of a gray area especially with regards to MIR semantics...

Reply:
Done. But I'd cc @RalfJung for their notice.

Reply:
Yeah this is way outside our current MIR semantics.^^ [See here](https://github.com/rust-lang/rfcs/pull/2850#issuecomment-581704306) for some discussion that could eventually be turned into a draft spec.